.lintstagedrc

```diff
@@ -0,0 +1,9 @@
+{
+    "**/*.+(ts|js)": [
+        "eslint"
+    ],
+    "**/*.+(ts|js|json|yml)": [
+        "prettier --write",
```

Comment:
So correct me if I am wrong the scenario is next:
- I commit changes
- It runs the pre-commit hook
- if there is a prettier error, it will fix it and add changes into stage

if this scenario happens, will it through an error and prevent committing?

Reply:
No, it won't, `prettier --write` returns nothing or 0, so it doesn't affect the process.

Reply:
I think it should prevent committing otherwise
- Developer commits changes
- It runs the pre-commit hook
- Developer ignores the error message
- Developers pushes invalid code
So what is the point to have pre-commit hook?

Maybe let's start from the beginning.
- Why do we need `pre-commit` hooks?
- What should it do?
- Do we need more hooks, such as `pre-push`?

From my perspective, the git hooks approach will help us find out the validity of the code before creating a PR automatically. Do we have more reasons to have it?
Maybe we need it only on pre-push. So we let developers commit possible invalid code but the developer is not able to Push this code? It means it should not overwrite User's code but throw an exception if it is invalid.

What do you think?

Reply:
> - Developer commits changes
> - It runs the pre-commit hook
> - Developer ignores the error message
> - Developers pushes invalid code

Why would developer ignore the error messages? If there are, then developer won't be able to commit. Maybe you misunderstood that only prettier errors will be fixed automatically? If there are any other errors related to ts, eslint it won't even get to the prettier part?

> if there is a prettier error, it will fix it and add changes into stage

You said it yourself, if there is a stylistic error, it will be fixed by prettier, so that there is no problem anymore, code is 100% valid.

> Maybe let's start from the beginning.

Let's

> - Why do we need pre-commit hooks?

Because we don't want to push invalid code to the remote. The code which would trigger CI and then fail anyway. Why to run non free CI if it may be cheaper and faster to check and eliminate on developer's PC.

> - What should it do?

- typechecking - developer should know if there is typescript issues. If there are, it should be eliminated.
- lint check - if there are lint related problems - those should be eliminated by developer.
- prettier - if there are stylistic issues, those would be fixed by prettier. There is no need to notify developer about wrong indent or missing brackets or semicolons if it can be fixed and pushed. Developer don't even need to care about styling rules on the project with this setup.
- build - the project should be buildable

> - Do we need more hooks, such as pre-push?

No, why? I don't see a reason for pre-push hook if all the checks were done on precommit hook.

> From my perspective, the git hooks approach will help us find out the validity of the code before creating a PR automatically. Do we have more reasons to have it?

What do you mean by "creating a PR automatically"? I mentioned above why we need hooks.

> Maybe we need it only on pre-push. So we let developers commit possible invalid code but the developers are not able to Push this code? It means it should not overwrite the code but throw an exception if it is invalid.

I don't think it's a good idea. Let's say developer commits 3 commits, and then he cannot push because there is an error related to the first commit. Then developer needs to fix it, and somehow append it to the first commit (by interactive rebase, or reseting / stashing or somehow else - a lot of hustle). In case hooks validation takes to long (the project is huge), then we would consider e.g. distributing tasks between different hooks, then we'd need to make compromises, but in case of our project, I don't see it becoming that big.

Reply:
> You said it yourself, if there is a stylistic error, it will be fixed by prettier, so that there is no problem anymore, code is valid.

yes but changes are not committed

> Because we don't want to push invalid code to the remote. The PR with errors would disturb other developers, trigger CI and then fail anyway.

but it does not solve it, it is still possible to push invalid code (prettier errors) with `git push`

> No, why? I don't see a reason for pre-push hook if all the checks were done on the pre-commit hook

yeah, but why to have a check on every commit instead of having just one before pushing?

I think it will get us slower, because us we mentioned above the main reason to have the hooks is to have a "clean" PR, not "clean" commits, I think `pre-push` will perfectly cover this part, for example
### with `pre-commit`
- I commit -> check ⚙️ -> it is valid ✅
- I commit -> check ⚙️ -> it is valid ✅
- I commit -> check ⚙️ -> it fails ⛔️
- (optional) -> I push -> Pipeline fails ⛔️⛔️
- I commit -> check ⚙️ -> it is valid ✅
- I push
### with `pre-push`
- I commit
- I commit
- I commit
- I push -> check ⚙️ -> it fails ⛔️
- I commit
- I push -> check ⚙️ -> it is valid ✅

So during an iteration, we will run 2 checks maximum -> the first to push and the second after fix

Reply:
> yes but changes are not committed

No! They are committed! Please try yourself, play with it and try different edge cases.

> but it does not solve it, it is still possible to push invalid code (prettier errors) with git push

No! It will be fixed automatically, and only then staged again. It's a _pre-commit_ hook, so everything what happens is _before_ commit. So it stages the fixed formatted files.

> yeah, but why to have a check on every commit instead of having just one before pushing?

I already described this case above 👇

> I don't think it's a good idea on our project. Let's say developer commits 3 commits, and then he cannot push because there is an error related to the first commit. Then developer needs to fix it, and somehow append it to the first commit (by interactive rebase, or reseting / stashing or somehow else - anyway it's a lot of hustle). In case hooks validation takes too long (the project is huge), then we would consider distributing tasks between different hooks, then we'd need to make compromises, but in case of our project, I don't see it becoming that big.

Please try it yourself first, it should answer your concerns.

Reply:
> with pre-push
> 1. I commit
> 1. I commit
> 1. I commit
> 1. I push -> check ⚙️ -> it fails ⛔️
> 1. I commit 
> 1. I push -> check ⚙️ -> it is valid ✅

On step 5 there is a problem if your mistake is related to the first commit. You would need to make merge commits that are 2 commits away, it's possible but would take longer than waiting few seconds more during commit. The hook is really a matter of few more seconds, please try to play with it and see.

Reply:
yeah, you are right, it commits. Agree, it changes my concerns. Let's merge it